### PR TITLE
Store double and float as std::string in the AST

### LINKED ISSUE #144: Possibility to store float, double as a string in the AST
If we have a floating point number in the AST, when we convert it back to NMODL, it would be usefult to keep the same representation/value as it was in the input.

For this, may be should store value of type string in the Float/Double node?

### PR DESCRIPTION
**Pull request by pramodk**
- when double/float values are parsed, stored in AST and printed
    back in the MOD file or C++ code, it could result into slightly
    different floating point value than the mod file
  - to avoid this and keep compatibility with neuron, we can store
    float/double as original parsed token string

fixes #144

WIP - there are still some things I would need to fix to pass all tests.

### CODE REVIEW

**Comment by cattabiani** [comment]
I report a cvf issue here as it is useful to decide how to tackle this problem:

File: src/visitors/units_visitor.cpp [hunk 1]
```diff
@@ -146,7 +147,7 @@ void UnitsVisitor::visit_factor_def(ast::FactorDef& node) {
         auto node_unit_name = node.get_node_name();
         auto unit1_factor = units_driver.table->get_unit(node_unit_name + "_unit1")->get_factor();
         auto unit2_factor = units_driver.table->get_unit(node_unit_name + "_unit2")->get_factor();
-        auto unit_factor = unit1_factor / unit2_factor;
+        auto unit_factor = stringutils::to_string(unit1_factor / unit2_factor);
```
Comment (iomaganaris) [inline src/visitors/units_visitor.cpp:150]: Here we should take care of the precision of the double converted to float according to the legacy or non legacy units similar to:
https://github.com/BlueBrain/nmodl/pull/421/files#diff-ecbcbc7e327becd3a246884e25ddd4d143b61bc35a7e39e68e4d5bf69d9b84eaR1997
Also the above should be edited to print a string

File: src/visitors/units_visitor.cpp [hunk 2]
```diff
@@ -146,7 +147,13 @@ void UnitsVisitor::visit_factor_def(ast::FactorDef& node) {
         auto node_unit_name = node.get_node_name();
         auto unit1_factor = units_driver.table->get_unit(node_unit_name + "_unit1")->get_factor();
         auto unit2_factor = units_driver.table->get_unit(node_unit_name + "_unit2")->get_factor();
-        auto unit_factor = unit1_factor / unit2_factor;
+
+#ifdef USE_LEGACY_UNITS
```
Comment (iomaganaris) [inline src/visitors/units_visitor.cpp:151]: Same as https://github.com/BlueBrain/mod2c/commit/e57ebb7b5ef2389d0782166056ed1afec1b94451#diff-ba4c77e010eaa2eac972c4e3e9412a19318b181cb6a8846858523e5120f4c7ff so all good 👍

**Review by iomaganaris — APPROVED** [review]
LGTM at this state.
Once the issue with the parsing of units is resolved then NMODL can be edited to follow the same procedure for reproducing the exact same values.

**Comment by pramodk** [comment]
@nrnhines : using your branch https://github.com/neuronsimulator/nrn/tree/dblunits in neuron, we get exact same values in neuron and nmodl ! :star: :clap:

```

    /** constants used in nmodl */
    static const double FARADAY1 = 96485.3321233100141;
    static const double FARADAY2 = 96.4853321233100161;
    static const double FARADAY3 = 9.64853321233100125;
    static const double PI = 3.14159265358979312;
    static const double R1 = 8.3144626181532395;
    static const double R2 = 8.314;
    static const double R3 = 8314.46261815323851;
    static const double R4 = 8.314;
    static const double R5 = 8.314500000000001;
    static const double dummy1 = 123.45;
    static const double dummy2 = 123.45e3;
    static const double dummy3 = 12345e-2;
    static const double KTOMV = 0.0853;
    static const double B = 0.26;
    static const double TEMP = 25;
    static const double FARADAY = 96.4853321233100161;
    static const double R = 8.313424;
```

vs neuron with `dblunits` branch:

```
#define FARADAY1 _nrnunit_FARADAY1[_nrnunit_use_legacy_]
static double _nrnunit_FARADAY1[2] = {0x1.78e555060882cp+16, 96485.3}; /* 96485.3321233100141 */

#define FARADAY2 _nrnunit_FARADAY2[_nrnunit_use_legacy_]
static double _nrnunit_FARADAY2[2] = {0x1.81f0fae775425p+6, 96.4853}; /* 96.4853321233100161 */

#define FARADAY3 _nrnunit_FARADAY3[_nrnunit_use_legacy_]
static double _nrnunit_FARADAY3[2] = {0x1.34c0c8b92a9b7p+3, 9.64853}; /* 9.64853321233100125 */

#define PI _nrnunit_PI[_nrnunit_use_legacy_]
static double _nrnunit_PI[2] = {0x1.921fb54442d18p+1, 3.14159}; /* 3.14159265358979312 */

#define R1 _nrnunit_R1[_nrnunit_use_legacy_]
static double _nrnunit_R1[2] = {0x1.0a1013e8990bep+3, 8.3145}; /* 8.3144626181532395 */
 static double R2 = 8.314;

#define R3 _nrnunit_R3[_nrnunit_use_legacy_]
static double _nrnunit_R3[2] = {0x1.03d3b37125759p+13, 8314.5}; /* 8314.46261815323851 */
 static double R4 = 8.314;
 static double R5 = 8.314500000000001;
 static double dummy1 = 123.45;
 static double dummy2 = 123.45e3;
 static double dummy3 = 12345e-2;
 static double KTOMV = 0.0853;
 static double B = 0.26;
 static double TEMP = 25.0;

#define FARADAY _nrnunit_FARADAY[_nrnunit_use_legacy_]
static double _nrnunit_FARADAY[2] = {0x1.81f0fae775425p+6, 96.4853}; /* 96.4853321233100161 */
```

And just for comparison, current master of neuron produces different values:

```
#define FARADAY1 _nrnunit_FARADAY1[_nrnunit_use_legacy_]
static double _nrnunit_FARADAY1[2] = {0x1.78e555060882dp+16, 96485.3}; /* 96485.3321233100287 */

#define FARADAY2 _nrnunit_FARADAY2[_nrnunit_use_legacy_]
static double _nrnunit_FARADAY2[2] = {0x1.81f0fae775426p+6, 96.4853}; /* 96.4853321233100303 */

#define FARADAY3 _nrnunit_FARADAY3[_nrnunit_use_legacy_]
static double _nrnunit_FARADAY3[2] = {0x1.34c0c8b92a9b8p+3, 9.64853}; /* 9.64853321233100303 */

#define PI _nrnunit_PI[_nrnunit_use_legacy_]
static double _nrnunit_PI[2] = {0x1.921fb54442d17p+1, 3.14159}; /* 3.14159265358979267 */

#define R1 _nrnunit_R1[_nrnunit_use_legacy_]
static double _nrnunit_R1[2] = {0x1.0a1013e8990bep+3, 8.3145}; /* 8.3144626181532395 */
 static double R2 = 8.314;

#define R3 _nrnunit_R3[_nrnunit_use_legacy_]
static double _nrnunit_R3[2] = {0x1.03d3b37125759p+13, 8314.5}; /* 8314.46261815323851 */
 static double R4 = 8.314;
 static double R5 = 8.314500000000001;
 static double dummy1 = 123.45;
 static double dummy2 = 123.45e3;
 static double dummy3 = 12345e-2;
 static double KTOMV = 0.0853;
 static double B = 0.26;
 static double TEMP = 25.0;

#define FARADAY _nrnunit_FARADAY[_nrnunit_use_legacy_]
static double _nrnunit_FARADAY[2] = {0x1.81f0fae775426p+6, 96.4853}; /* 96.4853321233100303 */
```

**Comment by pramodk** [comment]
With this PR, ISPC backend has same numbers as CPP backend:

```
/*********************************************************
Model Name      : cagk
Filename        : cagk.mod
NMODL Version   : 6.2.0
Vectorized      : true
Threadsafe      : true
Created         : Sat Dec  5 09:50:12 2020
Backend         : ispc (api-compatibility)
NMODL Compiler  : 0.2 [05d3e87 ]
*********************************************************/
#include "nmodl/fast_math.ispc"
#include "coreneuron/mechanism/nrnoc_ml.ispc"

/** constants used in nmodl */
static const uniform double FARADAY1 = 96485.3321233100141;
static const uniform double FARADAY2 = 96.4853321233100161;
static const uniform double FARADAY3 = 9.64853321233100125;
static const uniform double ISPC_PI = 3.14159265358979312;
static const uniform double R1 = 8.3144626181532395;
static const uniform double R2 = 8.314;
static const uniform double R3 = 8314.46261815323851;
static const uniform double R4 = 8.314;
static const uniform double R5 = 8.314500000000001;
static const uniform double dummy1 = 123.45;
static const uniform double dummy2 = 123.45e3;
static const uniform double dummy3 = 12345e-2;
static const uniform double KTOMV = 0.0853;
static const uniform double B = 0.26;
static const uniform double TEMP = 25;
static const uniform double FARADAY = 96.4853321233100161;
static const uniform double R = 8.313424;
```

So current fixes are sufficient for ISPC backend as well.

**Pull request merged**